### PR TITLE
Feature/ntlm redis cache

### PR DESCRIPTION
**Pull request by julsemaan**
# Description
Adds an NTLM authentication cache to FreeRADIUS via Redis
Reworks the domain page so modifying non-binding infos doesn't always trigger a join

# Impacts
NTLM authentication

# Delete branch after merge
YES

# NEWS file entries
## New Features
* Added an optional Redis driven cache for NTLM (Active Directory) 802.1x authentication

### CODE REVIEW

File: conf/documentation.conf [hunk 1]
```diff
@@ -320,6 +320,13 @@ description=<<EOT
 Should redis be started ? Keep enabled unless you know what you're doing.
 EOT
 
+[services.redis_ntlm_cache]
+type=toggle
+options=enabled|disabled
+description=<<EOT
+Should the redis NTLM cache be started ? Use this if you are enabling an Active Directory NTLM cache.
```
Comment (extrafu) [inline conf/documentation.conf]: redis => Redis
be started ?  => be started?

File: conf/documentation.conf [hunk 2]
```diff
@@ -1315,6 +1329,16 @@ description=<<EOT
...
 EOT
 
+[advanced.ntlm_redis_cache]
+type=toggle
+options=enabled|disabled
+description=<<EOT
+Enables a Redis driven cache for NTLM authentication. 
+In order for this to work, you need to setup proper NT hash syncronization between your PacketFence server and your AD.
+Refer to the Administration guide for more details.
+Applying this requires a restart of radiusd.
```
Comment (extrafu) [inline conf/documentation.conf:1370]: there's no "radiusd" - say "radius.authentication" - all of them? just be more precise.

Comment (julsemaan) [reply conf/documentation.conf:1370]: There is still a radiusd service, it represents all three of the services.

There currently no distinction between them when restarting the service

File: conf/pf.conf.defaults [hunk 3]
```diff
@@ -991,6 +1000,14 @@ hashing_cost=8
 # Disable Active-Directory domain configuration from within PacketFence
 disable_pf_domain_auth=disabled
 #
+# advanced.ntlm_redis_cache
+#
+# Enables a Redis driven cache for NTLM authentication. 
+# In order for this to work, you need to setup proper NT hash syncronization between your PacketFence server and your AD.
+# Refer to the Administration guide for more details.
+# Applying this requires a restart of radiusd.
```
Comment (extrafu) [inline conf/pf.conf.defaults:1033]: See above for comment

File: lib/pf/services/manager/radiusd_child.pm [hunk 4]
```diff
@@ -118,6 +118,25 @@ sub generate_radiusd_sitesconf {
         $tags{'multi_domain'} = '# packetfence-multi-domain not activated because no domains configured';
     }
 
+    if(isenabled($Config{advanced}{ntlm_redis_cache})) {
+        my $username_prefix = "NTHASH:%{%{PacketFence-Domain}:-''}";
+        $tags{'redis_ntlm_cache_fetch'} = <<EOT
+if(User-Name =~ /^host\\//) {
```
Comment (extrafu) [inline lib/pf/services/manager/radiusd_child.pm]: broken indent?

**Comment by julsemaan** [comment]
code is ready for review + to be tested

I'll follow up shortly with the proper admin documentation + load testing results

File: conf/documentation.conf [hunk 5]
```diff
@@ -303,28 +303,35 @@ EOT
...
 
 [services.redis_queue]
 type=toggle
 options=enabled|disabled
 description=<<EOT
-Should redis be started ? Keep enabled unless you know what you're doing.
+Should redis be started? Keep enabled unless you know what you're doing.
+EOT
+
+[services.redis_ntlm_cache]
```
Comment (jrouzierinverse) [inline conf/documentation.conf:332]: Where is the redis_ntlm_cache service manager?

Comment (julsemaan) [reply conf/documentation.conf:332]: At the most obvious place :P in the untracked files of my servers.

Pushing it right now

File: lib/pf/services/manager/redis_ntlm_cache.pm [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
+USA.
+
+=cut
+
+1;
```
Comment (rouzier) [inline lib/pf/services/manager/redis_ntlm_cache.pm:57]: Also the configuration file

Comment (julsemaan) [reply lib/pf/services/manager/redis_ntlm_cache.pm:57]: I'll let you guess where that was as well :)

**Comment by julsemaan** [comment]
Documentation completed + packaging for impacket done and placed in packetfence-extra for RHEL6 + 7

Currently still clueless around how Debian packaging works and there is no packetfence-extra on Debian. Will leave the decision on whether or not we distribute the packages to the PF Debian gods

File: lib/pf/domain/ntlm_cache.pm [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+=cut
+
+sub cache_user {
+    my ($domain, $username) = @_;
+    my $logger = get_logger;
+    my $config = $ConfigDomain{$domain};
+    my $source = getAuthenticationSource($config->{ntlm_cache_source});
+    return ($FALSE, "Invalid LDAP source $config->{ntlm_cache_source}") unless(defined($source));
+    
+    if($username =~ /^host\//) {
```
Comment (fdurand) [inline lib/pf/domain/ntlm_cache.pm:209]: Can we do something more generic ? Like the user attribute can be mail, cn ... so can we have something like userAttributeToSamAccountName instead ?

Comment (julsemaan) [reply lib/pf/domain/ntlm_cache.pm:209]: Just pushed something that should even exceed your expectation here.

Let me know if that works

**Comment by fdurand** [comment]
Yes i know but i already got a case with mail attribute (stuff in unlang in order to make it work).

**Comment by julsemaan** [comment]
Alright you win this one :P

**Comment by fdurand** [comment]
Rebase please

**Comment by julsemaan** [comment]
ribaized

File: lib/pf/domain/ntlm_cache.pm [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+
+use strict;
+use warnings;
+
+use pf::constants;
+use pf::authentication;
+use pf::config qw(%ConfigDomain);
+use Net::LDAP::Control::Paged;
+use Net::LDAP::Constant qw( LDAP_CONTROL_PAGED );
+use Data::Dumper;
```
Comment (fdurand) [inline lib/pf/domain/ntlm_cache.pm]: kindly remove it

File: lib/pf/domain/ntlm_cache.pm [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+    my ($connection, $LDAPServer, $LDAPServerPort ) = $ldap_source->_connect();
+
+    if (!defined($connection)) {
+        return ($FALSE, "Error communicating with the LDAP server");
+    }
+
+    my $page = Net::LDAP::Control::Paged->new(size => 1000) or die $!;
+    my @users;
+    my $cookie;
+    while(1) {
```
Comment (fdurand) [inline lib/pf/domain/ntlm_cache.pm]: $TRUE

File: lib/pf/domain/ntlm_cache.pm [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+}
+
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2016 Inverse inc.
```
Comment (fdurand) [inline lib/pf/domain/ntlm_cache.pm]: 2017

**Comment by fdurand** [comment]
IMO we have to merge it and test it

**Comment by julsemaan** [comment]
kindly re-review

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 11]
```diff
@@ -3863,6 +3863,62 @@ PacketFence can generate a lot of log entries in huge production environments. T
...
+
+NOTE: The NTLM cache doesn't cache clear text passwords, it caches the NT hash of the user password.
+
+Addtionnal packages
++++++++++++++++++++
+
+In order to use this cache, you will need to install additionnal packages available from the packetfence-extra repository. These packages will allow PacketFence to query your Active Directory for the NT hash of the users.
+
+```
+yum install python2-impacket --enablerepo=packetfence-extra
```
Comment (fdurand) [inline docs/PacketFence_Administration_Guide.asciidoc:3886]: apt-get install ....

File: html/pfappserver/root/config/domain/list.tt [hunk 12]
```diff
@@ -27,14 +27,26 @@
...
                 <div class="modal-header">
                   <a class="close" data-dismiss="modal">&times;</a>
-                  <h3>[% l('Please enter the password for the account :') %] [% item.bind_dn %]</h3>
+                  <h3>[% l('Please enter administrative credentials to connect to the domain') %]</h3>
                 </div>
 
                 <div class="modal-body">
-                    <input type="password" name="password" value="">
+                    <div class="control-group">
+                      <label class="control-label" for="bind_dn">Username</label>
```
Comment (fdurand) [inline html/pfappserver/root/config/domain/list.tt]: Localize please l('Username')

File: html/pfappserver/root/config/domain/list.tt [hunk 13]
```diff
@@ -27,14 +27,26 @@
...
+                      <div class="controls">
+                        <input type="text" name="username" value="">
+                      </div>
+                    </div>
+                    <br>
+                    <div class="control-group">
+                      <label class="control-label" for="bind_dn">Password</label>
```
Comment (fdurand) [inline html/pfappserver/root/config/domain/list.tt]: Localize please

File: html/pfappserver/root/config/domain/view.tt [hunk 14]
```diff
@@ -2,28 +2,32 @@
...
-        [% form.field('dns_name').render | none %]
-        [% form.field('server_name').render | none %]
-        [% form.field('ad_server').render | none %]
-        [% form.field('dns_server').render | none %]
-        [% form.field('bind_dn').render | none %]
-        [% form.field('bind_pass').render | none %]
-        [% form.field('ou').render | none %]
-        [% form.field('registration').render | none %]
+    <ul class="nav nav-tabs">
+      <li class="active" ><a href="#settings" data-toggle="tab">Settings</a></li>
```
Comment (fdurand) [inline html/pfappserver/root/config/domain/view.tt]: Localize please

File: html/pfappserver/root/config/domain/view.tt [hunk 15]
```diff
@@ -2,28 +2,32 @@
...
+      <li><a href="#ntlm-cache" data-toggle="tab">NTLM cache</a></li>
```
Comment (fdurand) [inline html/pfappserver/root/config/domain/view.tt]: Localize please

File: lib/pf/services/manager/redis_ntlm_cache.pm [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+has '+name' => (default => sub { 'redis_ntlm_cache' } );
+
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2016 Inverse inc.
```
Comment (fdurand) [inline lib/pf/services/manager/redis_ntlm_cache.pm]: 2017

File: docs/PacketFence_Administration_Guide.asciidoc (hunk 11 above)
Comment (julsemaan) [reply docs/PacketFence_Administration_Guide.asciidoc:3886]: kindly confirm this is good package: https://packages.debian.org/jessie/python-impacket

Comment (fdurand) [reply docs/PacketFence_Administration_Guide.asciidoc:3886]: Yes it is

**Comment by fdurand** [comment]
i kindly want to click on merge

**Comment by julsemaan** [comment]
fixed issues and internationalised strings
mr worldwide says hello

**Comment by julsemaan** [comment]
kindly click

**Pull request merged**